### ISSUE TITLE
capacity / stress test

### ISSUE THREAD

**Issue by ctm** [ctm/mb2-doc#231]
I believe mb2's core poker server can easily handle 100 times the load BARGE will put on it, however, there's more to a poker ecosystem than just the core server. 

I'm currently using an inexpensive ($30/month) RDS database instance and I'm not using a CDN. I store ridiculously large (for what they represent; they're not large in absolute size) JSON blobs in the database for logging, and although the client uses etags so it doesn't have to be constantly uploaded is a monolith that I don't even strip debug tags from. So ...

... I need to make sure that when we have a couple hundred people playing, nothing slows down. In fact, I really need to know at which points what sort of traffic will have what sort of performance considerations because there's a chance that mb2 will become exponentially popular as the UI improves and simple things like ring games and sit 'n' goes are added.

I have absolutely no experience provisioning CDNs. I have a bunch of (mostly antiquated) db knowledge, although never as a dba).

If I write the tools for testing 200 concurrent users, it should just be a parameter adjustment to also test for 1,000 or 10,000, although probably not for 100,000.

I should do this sooner rather than later. If there are surprises, I'd like to deal with them ASAP.

**Comment by ctm** [comment]
Look into how I'm using diesel. Compare that with what the [actix-web page recommends](https://actix.rs/docs/databases/).

**Comment by ctm** [comment]
Did a test with 40 w/o incident. Then did a test with 100 and ran into trouble. I suspect the trouble had to do more with my test harness, in that I multiplex all the tables through a single player and I _suspect_ that can result in weird session reuse that is impossible otherwise.

Specifically, during the 100 player test I wound up with a client crashing due to the nick being None. I also had mb2 crash with this panic:
```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', mb2/src/mb2/nick_mapper.rs:1109:66
stack backtrace:
   0:     0x561d869e26e5 - backtrace::backtrace::libunwind::trace::h34afbfad7fd770fc
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1:     0x561d869e26e5 - backtrace::backtrace::trace_unsynchronized::h460d522b1619a600
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2:     0x561d869e26e5 - std::sys_common::backtrace::_print_fmt::ha45fac10086813b4
                               at src/libstd/sys_common/backtrace.rs:78
   3:     0x561d869e26e5 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hde84f63fcfd0e6de
                               at src/libstd/sys_common/backtrace.rs:59
   4:     0x561d86a0aebc - core::fmt::write::h540ac4a6a1232abc
                               at src/libcore/fmt/mod.rs:1076
   5:     0x561d869dbe42 - std::io::Write::write_fmt::hc344eafd6e850b4d
                               at src/libstd/io/mod.rs:1537
   6:     0x561d869e4d70 - std::sys_common::backtrace::_print::h4db88ff15cb5d61d
                               at src/libstd/sys_common/backtrace.rs:62
   7:     0x561d869e4d70 - std::sys_common::backtrace::print::h5fc39e1b1f610bd3
                               at src/libstd/sys_common/backtrace.rs:49
   8:     0x561d869e4d70 - std::panicking::default_hook::{{closure}}::h59e55edacb1d974a
                               at src/libstd/panicking.rs:198
   9:     0x561d869e4abc - std::panicking::default_hook::heee4c8016dfbf328
                               at src/libstd/panicking.rs:217
  10:     0x561d869e53b3 - std::panicking::rust_panic_with_hook::h8405b6301c79fb5a
                               at src/libstd/panicking.rs:520
  11:     0x561d869e4fab - rust_begin_unwind
                               at src/libstd/panicking.rs:431
  12:     0x561d86a08b71 - core::panicking::panic_fmt::h78830ea6a34e7206
                               at src/libcore/panicking.rs:85
  13:     0x561d86a08abd - core::panicking::panic::h68e56c2eeba99c8c
                               at src/libcore/panicking.rs:50
  14:     0x561d86652e5f - core::option::Option<T>::unwrap::h99d28082fedccb19
                               at /rustc/16957bd4d3a5377263f76ed74c572aad8e4b7e59/src/libcore/macros/mod.rs:10
  15:     0x561d86652e5f - mb2::mb2::nick_mapper::NickMapper::moved_from_to::h56e7dab76c6229d1
                               at mb2/src/mb2/nick_mapper.rs:1109
```

It's too close to the evening tournament for me to continue, but I definitely need to do more on this. There should be no errors.

From a performance standpoint, I currently have all the players creating new window, so that and bandwidth into my house makes for a pretty big pause at the start. I'll look at the database numbers tomorrow.

Overall, I'm _fairly_ happy with how the stress testing is going. The downside is that time spent stress testing is time I'm not working on fun features.

**Comment by ctm** [comment]
Earlier this morning I ran a couple tournaments, one with 50 players connected to pop-up tables, the other with 200 players. The 50 player one had the default time-to-act of 45 seconds. For the 200 one, I lowered time-to-act to 5 seconds.

The load average during this test was 0.0. The RSS memory use was something like 65 MB.

I did run into a bug associated with quitting (#287) in a test even earlier, however I didn't quit any of the 250 players. I just let the blinds get big enough over the course of a couple hours so that eventually both tournaments completed.

While those two tournaments were going on, I fired up a demo tournament of 2 players and checked the responsiveness. Everything was snappy.

The above isn't a particularly grueling test, but it's enough for now.

My plan is to make a [few UI changes](https://github.com/ctm/mb2-doc/issues/289) and continue to do load testing off and on for the next few weeks, but under a new issue.
